src/frontend/src/handler/create_source.rs

```diff
@@ -797,42 +837,97 @@ pub(crate) fn bind_all_columns(
+                Some(col.name().to_string())
+            } else {
+                None
+            }
+        })
+        .collect_vec();
 
     let res = match (&source_schema.format, &source_schema.row_encode) {
         (Format::Native, Encode::Native) | (Format::Plain, _) => sql_defined_pk_names,
+
+        // For all Upsert formats, we only accept one and only key column as primary key.
+        // Additional KEY columns must be set in this case and must be primary key.
         (Format::Upsert, Encode::Json) => {
-            if sql_defined_pk {
+            if let Some(ref key_column_name) = key_column_name && sql_defined_pk {
+                if sql_defined_pk_names.len() != 1 {
+                    return Err(RwError::from(ProtocolError(
+                        format!("upsert json supports only one primary key column ({}).", key_column_name)
+                    )));
+                }
+                // the column name have been converted to real value in `handle_addition_columns`
+                // so we don't ignore ascii case here
+                if !key_column_name.eq(sql_defined_pk_names[0].as_str()) {
+                    return Err(RwError::from(ProtocolError(format!(
+                        "upsert json's key column {} not match with sql defined primary key {}",
+                        key_column_name, sql_defined_pk_names[0]
+                    ))));
+                }
                 sql_defined_pk_names
             } else {
-                add_default_key_column(columns);
-                vec![DEFAULT_KEY_COLUMN_NAME.into()]
+                return if key_column_name.is_none() {
+                    Err(
+                        RwError::from(ProtocolError("INCLUDE KEY clause must be set for FORMAT UPSERT ENCODE JSON".to_string()))
+                    )
+                } else {
+                    Err(RwError::from(ProtocolError(format!(
+                        "Primary key must be specified to {} when creating source with FORMAT UPSERT ENCODE JSON",
+                        key_column_name.unwrap()))))
+                }
             }
         }
         (Format::Upsert, Encode::Avro) => {
-            if sql_defined_pk {
-                if sql_defined_pk_names.len() != 1 {
-                    return Err(RwError::from(ProtocolError(
-                        "upsert avro supports only one primary key column.".to_string(),
-                    )));
+            // prev deprecated logic:
+            // if key schema can be inferred from schema registry, then use it
+            if sql_defined_pk && sql_defined_pk_names.len() != 1 {
+                return Err(RwError::from(ProtocolError(
+                    "upsert avro supports only one primary key column.".to_string(),
+                )));
+            }
+            if let Some(ref key_column_name) = key_column_name {
+                if key_column_name.eq(sql_defined_pk_names[0].as_str()) {
```

Comment:
> sql_defined_pk_names[0]

will panic if user does not define pk?

Reply:
merged with `format upsert encode json`, will check `sql_defined_pk_names` is not empty